Fix UInt128 to double conversion for values >= 2^104

sdcb opened the pull request:
Fixes #122203

sdcb committed:
Fix UInt128 to double conversion for values >= 2^104

Fixes #122203

sdcb committed:
Apply suggestion: change shift threshold to 40 and add test case

huoyaoyuan reviewed (commented):
Can you please check if the issue is also present for Int128?

sdcb committed:
Fix (U)Int128 to double conversion precision loss

- Correctly calculate the sticky bit when converting large UInt128 values (>= 2^104) to double to prevent precision loss.
- Optimize the upper bits extraction by accessing _upper directly instead of shifting.
- Add regression tests for Int128 to ensure it is also fixed.
- Update test comments to better explain the test cases.

Fixes #122203

sdcb commented:
@dotnet-policy-service agree

sdcb commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/System/UInt128Tests.cs:
```diff
@@ -517,5 +517,26 @@ public static void BigMul(UInt128 a, UInt128 b, string result)
+            // Value >= 2^104
+            // 2^104 = 20282409603651670423947251286016
+            // We add 2^24 + 1 to ensure we test the sticky bit logic if it matters,
+            // or at least that the large value path doesn't crash or produce garbage.
```
optimized

huoyaoyuan commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/System/UInt128Tests.cs:
```diff
@@ -517,5 +517,26 @@ public static void BigMul(UInt128 a, UInt128 b, string result)
+            // The value is constructed as 2^104 + 2^24 + 1.
+            // This tests a value with a 1 at bit 104, a 1 at bit 24, and a 1 at bit 0.
+            // The lower bits (24 and 0) should contribute to the sticky bit calculation.
+            UInt128 value2 = (UInt128.One << 104) + (UInt128.One << 24) + 1;
```
```suggestion
            UInt128 value2 = new(0x0100_0000_0000, 0x0100_0001);
```

huoyaoyuan reviewed (commented):
The numbers can be constructed more intuitively (also Int128 ones)

sdcb committed:
Make Int128/UInt128 test numbers more intuitive

Use decimal literals instead of scientific notation and hex constructors
instead of bit shifts for clearer test value representation.

sdcb committed:
Fix UInt128 to double conversion for values in [2^88, 2^104)

Change threshold check from (value._upper >> 24) to (value._upper >> 40)
to correctly handle values up to 2^104 in the intermediate precision branch.
